line_profiler/_logger.py

```diff
@@ -0,0 +1,286 @@
+"""
+# Ported from kwutil
+"""
+# import os
+import logging
+from abc import ABC, abstractmethod
+import sys
+from logging import INFO, DEBUG, ERROR, WARNING, CRITICAL  # NOQA
+
+
+class _LogBackend(ABC):
+    """
+    Abstract base class for our logger implementations.
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    @abstractmethod
+    def debug(self, msg, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def info(self, msg, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def warning(self, msg, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def error(self, msg, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def critical(self, msg, *args, **kwargs):
+        pass
+
+
+class _PrintLogBackend(_LogBackend):
+    """
+    A simple print-based logger that falls back to print output if no logging configuration
+    is set up.
+
+    Example:
+        >>> pl = _PrintLogBackend(name='print', level=INFO)
+        >>> pl.info('Hello %s', 'world')
+        Hello world
+        >>> pl.debug('Should not appear')
+    """
+
+    def __init__(self, name="<print-logger>", level=logging.INFO):
+        super().__init__(name)
+        self.level = level
+
+    def isEnabledFor(self, level):
+        return level >= self.level
+
+    def _log(self, level, msg, *args, **kwargs):
+        if self.isEnabledFor(level):
+            # Mimic logging formatting (ignoring extra kwargs for simplicity)
+            print(msg % args)
```

Comment:
Can we maybe do like `logging.Formatter` and have a `style` argument for switching between `%` and f-string interpolation? Like:
```suggestion
    def _log(self, level, msg, *args, style: "Literal['%', '{']" | None = None, **kwargs):
        if self.isEnabledFor(level):
            if style == '%':
                msg = msg % args
            elif style == '{':
                msg = msg.format(*args)
            # Mimic logging formatting (ignoring extra kwargs for simplicity)
            print(msg)
```

Reply:
This is mimicing the Python logging API, which ultimately calls:

```
    def getMessage(self):
        """
        Return the message for this LogRecord.

        Return the message for this LogRecord after merging any user-supplied
        arguments with the message.
        """
        msg = str(self.msg)
        if self.args:
            msg = msg % self.args
        return msg
```

The main reason I have any abstraction over it at all is because I often just want to get rid of python logging all together and use print statements. This is ported from a utility library (kwutil), and I'm also not 100% settled on anything in this yet. It might just make more sense to not include this here and just use stdlib logging normally.

Sorry for taking so long on all this. I would like to get your branch landed soon, but my time is spread thin right now. Hopefully within the month.

Reply:
Fair enough, I guess we can just keep it simple for now and add features/compatibility stuff as needed. Most of the time we do interpolations it suffices to just use `%s` and `%r` anyway...

It's alright, please take your time.